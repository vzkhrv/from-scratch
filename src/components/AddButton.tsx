import AddIcon from "../icons/add.png";

type Props = {
  onClick: () => void;
  disabled: boolean;
};
export const AddButton = ({ onClick, disabled }: Props) => {
  return (
    <button
      className="button button-with-icon"
      onClick={onClick}
      disabled={disabled}
      data-alt="добавить задачу"
    >
      <img src={AddIcon} alt="Добавить" />
    </button>
  );
};
