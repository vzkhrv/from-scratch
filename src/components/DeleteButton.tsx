import TrashIcon from "../icons/trash.png";

type Props = {
  onClick: () => void;
  disabled: boolean;
};
export const DeleteButton = ({ onClick, disabled }: Props) => {
  return (
    <button
      className="button button-with-icon"
      onClick={onClick}
      disabled={disabled}
      data-alt="Удалить"
    >
      <img src={TrashIcon} alt="Удалить" />
    </button>
  );
};
