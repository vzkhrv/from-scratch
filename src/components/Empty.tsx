export const Empty = () => {
  return (
    <div className="empty-wrapper">
      Вы пока не создали ни одной задачи. <br />Давайте что-нибудь запланируем?
    </div>
  );
};
