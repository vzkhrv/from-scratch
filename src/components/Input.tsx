import { ChangeEvent, useMemo } from "react";

import "./styles.css";
import { headerFieldOptions, validateHeaderMax } from "src/utils/helpers";

type Props = {
  value: string;
  onChange: (value: string) => void;
  disabledMessage?: string;
  disabled?: boolean;
};

export const Input = ({
  value,
  onChange,
  disabledMessage,
  disabled,
}: Props) => {
  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const val = event.target.value;

    onChange(val);
  };

  const hintValue = useMemo(
    () => (validateHeaderMax(value) ? "" : headerFieldOptions.message),
    [value]
  );

  return (
    <div className="input-field-wrapper" data-testid="input-container">
      <input
        id="input-header-field-id"
        type="text"
        className="input-field-element input-field-node input-field-input"
        value={value}
        onChange={handleChange}
        placeholder={
          disabled ? disabledMessage : `например, введите "купить молоко"`
        }
        alt="поле для ввода заголовка задачи"
        title="поле для заголовка"
        data-testid="input-field"
        style={{ display: "flex" }}
        disabled={disabled}
      />
      <span data-testid="input-hint-text" className="input-field-hint">
        {hintValue}
      </span>
    </div>
  );
};
