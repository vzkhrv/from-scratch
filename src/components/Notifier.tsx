import { useEffect } from "react";

type Props = {
    task: string;
    open: boolean;
    onClose: () => void;
}

export const Notifier = ({ task, open, onClose }: Props) => {
    useEffect(() => {
        let timerId: NodeJS.Timeout;

        if (open) {
            timerId = setTimeout(() => {
                onClose();
            }, 2000);
        }

        return () => clearTimeout(timerId);
    }, [open]);

    if (!open) {
        return null;
    }

    return(
        <div className="blackout">
            <div className="notifier-wrapper">
                {task}
            </div>
        </div>
    );
}