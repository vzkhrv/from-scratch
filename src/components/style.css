.input-field {
    display: flex;
    flex-flow: column nowrap;

    width: 300px;
    padding: 24px;
}

.input-field label {
    font-size: 10px;
    color: gray;
}