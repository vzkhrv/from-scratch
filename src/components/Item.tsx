import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
      />
      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {props.done ? <s>{props.header}</s> : props.header}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
