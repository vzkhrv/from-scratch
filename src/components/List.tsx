import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => (
  <ul className="task-list tasks">
    {items.map((item) => (
      <Item
        {...item}
        key={item.id}
        onDelete={onDelete}
        onToggle={onToggle} />
    ))}
  </ul>
);
