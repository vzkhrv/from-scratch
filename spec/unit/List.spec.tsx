import { render, screen } from "@testing-library/react";
import { List } from "src/components/List";

it("отображение списка задач", () => {
  const onDelete = jest.fn();
  const onToggle = jest.fn();

  const items: Task[] = [
    {
      id: "1",
      header: "купить хлеб",
      done: false,
    },
    {
      id: "2",
      header: "купить молоко",
      done: false,
    },
    {
      id: "3",
      header: "выгулять собаку",
      done: true,
    },
  ];

  const { rerender, asFragment } = render(
    <List items={items} onDelete={onDelete} onToggle={onToggle} />
  );
  const firstRender = asFragment();
  
  items.pop();

  rerender(<List items={items} onDelete={onDelete} onToggle={onToggle} />);
  const secondRender = asFragment();

  expect(firstRender).toMatchDiffSnapshot(secondRender);
});

it("Список содержит не больше 10 невыполненных задач", () => {

});
