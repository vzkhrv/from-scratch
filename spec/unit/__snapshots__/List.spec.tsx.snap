// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`отображение списка задач 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -46,32 +46,7 @@
            alt="Удалить"
            src="[object Object]"
          />
        </button>
      </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         checked=""
-         id="3"
-         type="checkbox"
-       />
-       <label
-         for="3"
-       >
-         <s>
-           выгулять собаку
-         </s>
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
    </ul>
  </DocumentFragment>"
`;
